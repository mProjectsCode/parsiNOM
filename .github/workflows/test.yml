name: Test

on:
  push:
    branches: [master, main]
  pull_request:
    branches: [master, main]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest
      - working-directory: .
        run: bun install
      - working-directory: .
        run: bun run check
