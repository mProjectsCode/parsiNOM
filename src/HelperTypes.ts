import type { Parser } from './Parser';
import type { ParserContext } from './ParserContext';

export type STypeBase = unknown;

export interface ParseSuccess<SType extends STypeBase> {
	success: true;
	value: SType;
	furthest?: undefined;
	expected?: undefined;
}

export interface ParseFailure {
	success: false;
	value: undefined;
	furthest: number;
	expected: string[];
}

export type ParseResult<SType extends STypeBase> = ParseSuccess<SType> | ParseFailure;

export interface InternalParseFailure {
	success: false;
	value: undefined;
}

export type InternalParseResult<SType extends STypeBase> = ParseSuccess<SType> | InternalParseFailure;

export type ParseFunction<SType extends STypeBase> = (context: ParserContext) => InternalParseResult<SType>;

export interface ParsingRange {
	from: number;
	to: number;
}

export interface ParsingMarker<SType extends STypeBase> {
	value: SType;
	range: ParsingRange;
}

export interface NamedParsingMarker<SType extends STypeBase> {
	value: SType;
	name: string;
	range: ParsingRange;
}

export type LanguageDef<RuleNames extends object> = {
	[P in keyof RuleNames]: Parser<RuleNames[P]>;
};

export type LanguageRules<RuleNames extends object> = {
	[P in keyof RuleNames]: (language: LanguageDef<RuleNames>) => Parser<RuleNames[P]>;
};

/**
 * All parsers of a language. Returned by {@link P.createLanguage}.
 */
export type NomLanguage<RulesType extends object> = {
	[P in keyof RulesType]: Parser<RulesType[P]>;
};

/**
 * All parsers of a language. Returned by {@link P.createLanguage}.
 */
export type NomLanguageRules<RulesType extends object> = {
	readonly [P in keyof RulesType]: (language: NomLanguagePartial<RulesType>, ref: NomLanguageRef<RulesType>) => Parser<RulesType[P]>;
};

/**
 * All parsers of a language wrapped in {@link P.reference}.
 */
export type NomLanguageRef<RulesType extends object> = {
	[P in keyof RulesType]: ParserRef<RulesType[P]>;
};

/**
 * Parsers of a language that came before this parser. All other parsers will be undefined.
 */
export type NomLanguagePartial<RulesType extends object> = {
	[P in keyof RulesType]: Parser<RulesType[P]>;
};

export type ParserRef<SType extends STypeBase> = Parser<SType>;

export type TupleToUnion<T extends readonly unknown[]> = T[number];

export type DeParser<T extends Parser<unknown>> = T extends Parser<infer U> ? U : never;

export type DeParserArray<T extends readonly Parser<unknown>[]> = {
	[I in keyof T]: T[I] extends Parser<infer R> ? R : never;
};
