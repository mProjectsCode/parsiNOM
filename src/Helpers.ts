import { type ParseResult, type ParsingPosition, type STypeBase } from './HelperTypes';
import { Parser } from './Parser';

export function arrayUnion(a: string[] | undefined, b: string[] | undefined): string[] | undefined {
	if (a === undefined && b === undefined) {
		return undefined;
	}
	if (a === undefined) {
		return b;
	}
	if (b === undefined) {
		return a;
	}

	for (const bElement of b) {
		if (!a.includes(bElement)) {
			a.push(bElement);
		}
	}
	return a;
}

export function getIndex(position: ParsingPosition | undefined): number {
	return position === undefined ? -1 : position.index;
}

export function validateRange(min: number, max: number): void {
	if (max < min) {
		throw new Error(`Invalid Range: max might not be smaller than min. Received [${min}, ${max}].`);
	}

	if (min < 0 || max < 0) {
		throw new Error(`Invalid Range: max and min might not be smaller than 0. Received [${min}, ${max}].`);
	}

	if (min === Infinity || max === Infinity) {
		throw new Error(`Invalid Range: max and min might not be infinity. Received [${min}, ${max}].`);
	}

	if (!Number.isInteger(min) || !Number.isInteger(max)) {
		throw new Error(`Invalid Range: max and min must be integers. Received [${min}, ${max}].`);
	}
}

export function validateRegexFlags(flags: string): void {
	for (const flag of flags) {
		if (flag !== 'i' && flag !== 'm' && flag !== 's' && flag !== 'u') {
			throw new Error(`RegExp flag '${flag}' is not allowed. The only allowed flags are 'i', 'm', 's' and 'u'.`);
		}
	}
}

// Parser Helpers

export class ParserHelpers {
	followedBy<SType extends STypeBase>(x: Parser<SType>): Parser<SType> {
		return new Parser<SType>(function _followedBy(context): ParseResult<SType> {
			return x.p(context.copy());
		});
	}

	/**
	 * Returns a parser that yields undefined if the input parser fails, and fails if the input parser accepts. Consumes no input.
	 * More or less an inverse lookahead.
	 *
	 * @param parser
	 */
	notFollowedBy(parser: Parser<unknown>): Parser<undefined> {
		return new Parser(function _notFollowedBy(context): ParseResult<undefined> {
			const contextCopy = context.copy();
			const result = parser.p(contextCopy);
			const text = context.sliceTo(contextCopy.position.index);
			return result.success ? context.fail(`not '` + text + `'`) : context.succeed(undefined);
		});
	}

	/**
	 * Returns a parser that passes the next character into `fn` and yields the character if `fn` returns true.
	 *
	 * @param fn
	 */
	test(fn: (char: string) => boolean): Parser<string> {
		return new Parser<string>(function _test(context): ParseResult<string> {
			const char = context.input[context.position.index];
			if (!context.atEOF() && fn(char)) {
				return context.succeedOffset(1, char);
			} else {
				return context.fail(`a character matching ${fn}`);
			}
		});
	}
}

export const P_HELPERS = new ParserHelpers();
