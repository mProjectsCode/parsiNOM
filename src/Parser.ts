import { P_HELPERS, validateRange } from './Helpers';
import type { NamedParsingMarker, ParseFunction, ParseResult, ParsingMarker, ParsingRange, STypeBase } from './HelperTypes';
import { ParserContext } from './ParserContext';
import { ParsingError } from './ParserError';
import { P_UTILS } from './ParserUtils';
import { P } from './ParsiNOM';

export class Parser<const SType extends STypeBase> {
	public p: ParseFunction<SType>;

	constructor(p: ParseFunction<SType>) {
		this.p = p;
	}

	/**
	 * Parses a string, returning a result object.
	 *
	 * @param str
	 */
	tryParse(str: string): ParseResult<SType> {
		const context = new ParserContext(str, 0);
		const result = this.p(context);
		if (result.success) {
			return result;
		} else {
			return {
				success: false,
				value: undefined,
				furthest: context.latestError!.position,
				expected: context.latestError!.expected,
			};
		}
	}

	/**
	 * Parses a string, throwing a {@link ParsingError} on failure.
	 *
	 * @param str
	 */
	parse(str: string): SType {
		const result: ParseResult<SType> = this.tryParse(str);
		if (result.success) {
			return result.value;
		} else {
			throw new ParsingError(str, result);
		}
	}

	/**
	 * Tries this parser first and then the second if this one fails.
	 *
	 * @param other
	 */
	or<OtherSType extends STypeBase>(other: Parser<OtherSType>): Parser<SType | OtherSType> {
		return P.or(this as Parser<SType>, other);
	}

	/**
	 * Wrap this parser with the same parser on both sides.
	 * `a.trim(b)` is the same as `a.wrap(b, b)`.
	 *
	 * @param parser
	 */
	trim(parser: Parser<unknown>): Parser<SType> {
		return this.wrap(parser, parser);
	}

	/**
	 * Wrap this parser with the same string on both sides.
	 * `a.trimString(str)` is the same as `a.trim(P.string(str))` or `a.wrap(P.string(str), P.string(str))` or `a.wrapString(str, str)`.
	 *
	 * @param str
	 */
	trimString(str: string): Parser<SType> {
		return this.trim(P.string(str));
	}

	/**
	 * Wrap this parser with two parsers, one on each side.
	 * `parser.wrap(leftParser, rightParser)` is the same as `leftParser.then(parser).skip(rightParser)` or `P.sequenceMap((left, middle, right) => middle, leftParser, parser, rightParser)`.
	 *
	 * @param leftParser
	 * @param rightParser
	 */
	wrap(leftParser: Parser<unknown>, rightParser: Parser<unknown>): Parser<SType> {
		const _this = this;

		return new Parser<SType>(function _wrap(context) {
			const leftResult = leftParser.p(context);
			if (!leftResult.success) {
				return leftResult;
			}

			const thisResult = _this.p(context);
			if (!thisResult.success) {
				return thisResult;
			}

			const rightResult = rightParser.p(context);
			if (!rightResult.success) {
				return rightResult;
			}

			return thisResult;
		});
	}

	/**
	 * Wrap this parser with two strings, one on each side.
	 * `parser.wrapString(leftStr, rightStr)` is the same as `parser.wrap(P.string(leftStr), P.string(rightStr))`.
	 *
	 * @param leftStr
	 * @param rightStr
	 */
	wrapString(leftStr: string, rightStr: string): Parser<SType> {
		return this.wrap(P.string(leftStr), P.string(rightStr));
	}

	/**
	 * Follow this parser with another parser, but the value of the other parser is returned.
	 * Similar to {@link Parser.skip}.
	 *
	 * @param next
	 */
	then<OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<OtherSType> {
		const _this = this;

		return new Parser<OtherSType>(function _then(context) {
			const firstResult = _this.p(context);
			if (!firstResult.success) {
				return firstResult;
			}

			const secondResult = next.p(context);
			if (!secondResult.success) {
				return secondResult;
			}

			return secondResult;
		});
	}

	/**
	 * Follow this parser with another parser, but the value of this parser is returned.
	 * Similar to {@link Parser.then}.
	 *
	 * @param next
	 */
	skip(next: Parser<unknown>): Parser<SType> {
		const _this = this;

		return new Parser<SType>(function _skip(context) {
			const firstResult = _this.p(context);
			if (!firstResult.success) {
				return firstResult;
			}

			const secondResult = next.p(context);
			if (!secondResult.success) {
				return secondResult;
			}

			return firstResult;
		});
	}

	/**
	 * Follow this parser with another parser.
	 * If you chain more than two parser use {@link P.sequence}.
	 *
	 * @param next
	 */
	and<OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<[SType, OtherSType]> {
		return P.sequence(this as Parser<SType>, next);
	}

	/**
	 * Matches this parser as often as it can. Potentially zero or infinite times.
	 */
	many(): Parser<SType[]> {
		const _this = this;
		return new Parser<SType[]>(function _many(context): ParseResult<SType[]> {
			const value: SType[] = [];

			while (true) {
				const beforeIndex = context.position;
				const result = _this.p(context);

				if (result.success) {
					if (context.position === beforeIndex) {
						throw new Error('infinite loop in many() parser detected');
					}

					value.push(result.value);
				} else {
					context.position = beforeIndex;

					return {
						success: true,
						value: value,
					};
				}
			}
		});
	}

	/**
	 * Match this parser at least `min` and at most `max` times.
	 * If you need only an upper or lower bound see {@link Parser.atLeast} or {@link Parser.atMost}.
	 *
	 * @param min
	 * @param max
	 */
	repeat(min: number, max: number): Parser<SType[]> {
		validateRange(min, max);
		const _this = this;

		return new Parser<SType[]>(function _repeat(context) {
			const value: SType[] = [];
			let count = 0;

			while (count < max) {
				const beforeIndex = context.position;
				const result = _this.p(context);

				if (result.success) {
					if (context.position === beforeIndex) {
						throw new Error('infinite loop in many() parser detected');
					}

					value.push(result.value);
					count++;
				} else {
					context.position = beforeIndex;

					if (count >= min) {
						return {
							success: true,
							value: value,
						};
					} else {
						return {
							success: false,
							value: undefined,
						};
					}
				}
			}

			return {
				success: true,
				value: value,
			};
		});
	}

	/**
	 * Match this parser at most `max` times.
	 *
	 * @param max
	 */
	atMost(max: number): Parser<SType[]> {
		return this.repeat(0, max);
	}

	/**
	 * Match this parser at least `min` times.
	 *
	 * @param min
	 */
	atLeast(min: number): Parser<SType[]> {
		return P.sequenceMap(
			(part1, part2) => {
				return part1.concat(part2);
			},
			this.repeat(min, min),
			this.many(),
		);
	}

	/**
	 * Same as {@link Parser.separateByNotEmpty}, but it also accepts empty inputs.
	 *
	 * @param separator
	 */
	separateBy(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateBy(this, separator);
	}

	/**
	 * Matches this parser multiple times, separated by some other parser, e.g. comma seperated values. Does not accept empty input.
	 *
	 * @param separator
	 */
	separateByNotEmpty(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateByNotEmpty(this, separator);
	}

	/**
	 * Change the result of the parser.
	 *
	 * @example A parser that matches the string 'true' but yields the boolean value 'true'.
	 * P.string('true').result(true)
	 *
	 * @param value
	 */
	result<OtherSType extends STypeBase>(value: OtherSType): Parser<OtherSType> {
		return this.map(() => value);
	}

	/**
	 * Allows for transformation of the return value after the parser has run.
	 *
	 * @param fn
	 */
	map<OtherSType extends STypeBase>(fn: (value: SType) => OtherSType): Parser<OtherSType> {
		const _this = this;
		return new Parser(function _map(context) {
			// we use any here, because that allows us to change the value of the result later
			const result = _this.p(context);
			if (!result.success) {
				return result;
			}

			return {
				success: true,
				value: fn(result.value),
			};
		});
	}

	/**
	 * Wrap the return value if this parser in an object containing the before and after parsing position.
	 */
	marker(): Parser<ParsingMarker<SType>> {
		return P.sequenceMap(
			function _marker(from, value: SType, to): ParsingMarker<SType> {
				return {
					value: value,
					range: { from, to },
				};
			},
			P_UTILS.position(),
			this as Parser<SType>,
			P_UTILS.position(),
		);
	}

	/**
	 * Same as {@link Parser.marker} but allows for naming of the marker.
	 *
	 * @param name
	 */
	namedMarker(name: string): Parser<NamedParsingMarker<SType>> {
		return P.sequenceMap(
			function _namedMarker(from, value: SType, to): NamedParsingMarker<SType> {
				return {
					value: value,
					name: name,
					range: { from, to },
				};
			},
			P_UTILS.position(),
			this as Parser<SType>,
			P_UTILS.position(),
		);
	}

	/**
	 * Similar to {@link Parser.map}, but with additional access to a parsing range.
	 *
	 * @param fn
	 */
	node<NodeType>(fn: (value: SType, range: ParsingRange) => NodeType): Parser<NodeType> {
		return P.sequenceMap(
			function _node(from, value: SType, to): NodeType {
				return fn(value, { from, to });
			},
			P_UTILS.position(),
			this as Parser<SType>,
			P_UTILS.position(),
		);
	}

	/**
	 * Functions like lookahead. Checks if this parser is followed by `next`, but does not advance the parsing position.
	 * So it is similar to {@link Parser.skip} but it does not advance the parsing position.
	 *
	 * @param next
	 */
	followedBy(next: Parser<unknown>): Parser<SType> {
		return this.skip(P_HELPERS.followedBy(next));
	}

	/**
	 * Functions like inverse lookahead. Checks if this parser is **not** followed by `next`, but does not advance the parsing position.
	 * So it is more or less the inverse of {@link Parser.followedBy}.
	 *
	 * @param next
	 */
	notFollowedBy(next: Parser<unknown>): Parser<SType> {
		return this.skip(P_HELPERS.notFollowedBy(next));
	}

	/**
	 * Overrides the error message of this parser when it fails.
	 *
	 * @param expected
	 */
	describe(expected: string | string[]): Parser<SType> {
		if (typeof expected === 'string') {
			expected = [expected];
		}

		const _this = this;

		return new Parser<SType>(function _describe(context) {
			const oldLatestError = context.getAndClearLatestError();

			const result = _this.p(context);

			if (context.latestError !== undefined) {
				context.latestError.expected = expected;
			}

			context.mergeLatestError(oldLatestError);

			return result;
		});
	}

	/**
	 * Aggregates the error messages of the parser and gives them an overarching description.
	 *
	 * a or b => (a or b as part of `expected`)
	 *
	 * @param expected
	 */
	box(expected: string): Parser<SType> {
		const _this = this;

		return new Parser<SType>(function _describe(context) {
			const oldLatestError = context.getAndClearLatestError();

			const result = _this.p(context);

			if (context.latestError !== undefined) {
				context.latestError.expected = [`(${context.latestError.expected.join(' or ')} as part of ${expected})`];
			}

			context.mergeLatestError(oldLatestError);

			return result;
		});
	}

	/**
	 * Makes the parser optional. Provide an optional fallback value that will be yielded when the parser fails, otherwise the parser yields `undefined`, but it will never fail.
	 */
	optional(): Parser<SType | undefined>;
	optional<OtherSType extends STypeBase>(value: OtherSType): Parser<SType | OtherSType>;
	optional<OtherSType extends STypeBase>(value?: OtherSType): Parser<SType | OtherSType | undefined> {
		return this.or(P.succeed(value));
	}

	/**
	 * Specify a function that returns a followup parser based on the result of this parser.
	 *
	 * @param fn
	 */
	chain<OtherSType extends STypeBase>(fn: (result: SType) => Parser<OtherSType>): Parser<OtherSType> {
		const _this = this;

		return new Parser<OtherSType>(function _chain(context) {
			const result = _this.p(context);
			if (!result.success) {
				return result;
			}
			return fn(result.value).p(context);
		});
	}

	/**
	 * Makes this parser expect that it is at the end of the input after parsing.
	 */
	thenEof(): Parser<SType> {
		const _this = this;

		return new Parser<SType>(function _thenEof(context) {
			const result = _this.p(context);
			if (!result.success) {
				return result;
			}
			if (!context.atEOF()) {
				return context.fail('eof');
			}
			return result;
		});
	}
}
