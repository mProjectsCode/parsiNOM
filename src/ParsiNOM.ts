import { P_HELPERS, validateRegexFlags } from './Helpers';
import type {
	DeParserArray,
	InternalParseResult,
	NomLanguage,
	NomLanguagePartial,
	NomLanguageRef,
	NomLanguageRules,
	ParseFunction,
	ParserRef,
	STypeBase,
	TupleToUnion,
} from './HelperTypes';
import { Parser } from './Parser';

export class P {
	// --- COMBINATORS ---

	/**
	 * Matches multiple parsers in a row, yielding an array of their results.
	 *
	 * @param parsers
	 */
	static sequence<ParserArr extends readonly Parser<unknown>[]>(...parsers: ParserArr): Parser<DeParserArray<ParserArr>> {
		if (parsers.length === 0) {
			throw new Error('sequence must have at least one parser argument');
		}

		return new Parser<DeParserArray<ParserArr>>(function _sequence(context): InternalParseResult<DeParserArray<ParserArr>> {
			const value: unknown[] = new Array(parsers.length);

			for (let i = 0; i < parsers.length; i++) {
				const p = parsers[i];

				const result = p.p(context);
				if (!result.success) {
					return result;
				}

				value[i] = result.value;
			}

			return {
				success: true,
				value: value as DeParserArray<ParserArr>,
			};
		});
	}

	/**
	 * Matches multiple parsers in a row, passing the result into `fn` and yielding the return value of `fn`.
	 *
	 * @param fn
	 * @param parsers
	 */
	static sequenceMap<OtherSType extends STypeBase, ParserArr extends readonly Parser<unknown>[]>(
		fn: (...value: DeParserArray<ParserArr>) => OtherSType,
		...parsers: ParserArr
	): Parser<OtherSType> {
		if (parsers.length === 0) {
			throw new Error('sequenceMap must have at least one parser argument');
		}

		return new Parser<OtherSType>(function _sequenceMap(context): InternalParseResult<OtherSType> {
			const value: unknown[] = new Array(parsers.length);

			for (let i = 0; i < parsers.length; i++) {
				const p = parsers[i];

				const result = p.p(context);
				if (!result.success) {
					return result;
				}

				value[i] = result.value;
			}

			return {
				success: true,
				value: fn(...(value as DeParserArray<ParserArr>)),
			};
		});
	}

	/**
	 * Utility for creating languages.
	 * Use `language` to refer to rules defined previously and `ref` to refer to rules that are defined later and the same rule.
	 * You can also use `ref` to refer to all other rules, but that will be slower.
	 *
	 * @param parsers
	 */
	static createLanguage<const RulesType extends object>(parsers: NomLanguageRules<RulesType>): NomLanguage<RulesType> {
		const language: NomLanguage<RulesType> = {} as NomLanguage<RulesType>;
		const languageProxy: NomLanguagePartial<RulesType> = {} as NomLanguagePartial<RulesType>;
		const languageRef: NomLanguageRef<RulesType> = {} as NomLanguageRef<RulesType>;

		for (const key in parsers) {
			languageRef[key] = P.reference(() => parsers[key](language, languageRef));
			// use getters on language proxy, so that we can throw errors when accessing later rules.
			Object.defineProperty(languageProxy, key, {
				get: () => {
					if (language[key] !== undefined) {
						return language[key];
					} else {
						throw new Error(`Can not access rule '${key}' in language. Rule is not yet defined. Try to access it via 'ref'.`);
					}
				},
			});
		}

		for (const key in parsers) {
			language[key] = parsers[key](languageProxy, languageRef);
		}

		return language;
	}

	/**
	 * Takes in a list of parsers and tries them in order until one succeeds, then returns that parsers result.
	 *
	 * @param parsers
	 */
	static or<ParserArr extends readonly Parser<unknown>[]>(...parsers: ParserArr): Parser<TupleToUnion<DeParserArray<ParserArr>>> {
		if (parsers.length === 0) {
			throw new Error('or must have at least one alternative');
		}

		return new Parser<TupleToUnion<DeParserArray<ParserArr>>>(function _or(context): InternalParseResult<TupleToUnion<DeParserArray<ParserArr>>> {
			const startPosition = context.position;

			for (const parser of parsers) {
				const p = parser as Parser<TupleToUnion<DeParserArray<ParserArr>>>;

				context.position = startPosition;
				const result = p.p(context);

				if (result.success) {
					return result;
				}
			}

			return {
				success: false,
				value: undefined,
			};
		});
	}

	/**
	 * Same as {@link P.separateByNotEmpty}, but it also accepts empty inputs.
	 *
	 * @param parser
	 * @param separator
	 */
	static separateBy<SType extends STypeBase>(parser: Parser<SType>, separator: Parser<unknown>): Parser<SType[]> {
		return this.separateByNotEmpty(parser, separator).or(P.succeed([]));
	}

	/**
	 * Matches a separated list, so e.g. comma separated values. Yields the values as an array. Does not accept empty input.
	 *
	 * @param parser
	 * @param separator
	 */
	static separateByNotEmpty<SType extends STypeBase>(parser: Parser<SType>, separator: Parser<unknown>): Parser<SType[]> {
		return this.sequenceMap(
			function _separateBy(part1, part2): SType[] {
				// console.log('sep', [part1, ...part2]);
				part2.unshift(part1);
				return part2;
			},
			parser,
			separator.then(parser).many(),
		);
	}

	// --- CONSTRUCTORS ---

	/**
	 * Matches a string. Yields the string.
	 *
	 * @param str
	 */
	static string(str: string): Parser<string> {
		const expected = "'" + str + "'";

		return new Parser<string>(function _string(context): InternalParseResult<string> {
			for (let i = 0; i < str.length; i++) {
				if (context.input[context.position + i] !== str[i]) {
					return context.fail(expected);
				}
			}

			return context.succeedOffset(str.length, str);
		});
	}

	/**
	 * Matches a regexp. Yields the matched string or the content of a specified capture group.
	 *
	 * @param regexp
	 * @param group
	 */
	static regexp(regexp: RegExp, group?: number): Parser<string> {
		validateRegexFlags(regexp.flags);

		const expected = regexp.source;

		if (group !== undefined) {
			return new Parser<string>(function _regexp(context): InternalParseResult<string> {
				const subInput = context.input.slice(context.position);
				const match = regexp.exec(subInput);

				if (match !== null) {
					const captureGroup = group ?? 0;

					if (captureGroup >= 0 && captureGroup <= match.length) {
						const fullMatch = match[0];
						const groupMatch = match[captureGroup];
						// console.log('regexp', expected, fullMatch, context);
						return context.succeedOffset(fullMatch.length, groupMatch);
					}

					const message = 'expected valid match group (0 to ' + match.length + ') in ' + expected;
					return context.fail(message);
				} else {
					return context.fail(expected);
				}
			});
		} else {
			return new Parser<string>(function _regexp(context): InternalParseResult<string> {
				const subInput = context.input.slice(context.position);
				const match = regexp.exec(subInput);

				if (match !== null) {
					const fullMatch = match[0];
					return context.succeedOffset(fullMatch.length, fullMatch);
				} else {
					return context.fail(expected);
				}
			});
		}
	}

	/**
	 * Parser that always succeeds and yields the value passed in as the argument. Consumes no input.
	 *
	 * @param value
	 */
	static succeed<SType extends STypeBase>(value: SType): Parser<SType> {
		return new Parser<SType>(function _succeed(context): InternalParseResult<SType> {
			return context.succeed(value);
		});
	}

	/**
	 * Parser that always fails and expects the value provided. Consumes no input.
	 *
	 * @param expected
	 */
	static fail<SType extends STypeBase>(expected: string): Parser<SType> {
		return new Parser<SType>(function _fail(context): InternalParseResult<SType> {
			return context.fail(expected);
		});
	}

	/**
	 * Returns a parser that yields the next character if it is included in `str`.
	 * If you want to match multiple characters in a row consider using {@link P.manyOf}.
	 *
	 * @param str
	 */
	static oneOf(str: string): Parser<string> {
		return P_HELPERS.test(function _oneOf(char: string): boolean {
			return str.includes(char);
		}, `one character of '${str}'`);
	}

	/**
	 * Returns a parser that yields the next character if it is **not** included in `str`.
	 * If you want to match multiple characters in a row consider using {@link P.manyNotOf}.
	 *
	 * @param str
	 */
	static noneOf(str: string): Parser<string> {
		return P_HELPERS.test(function _noneOf(char: string): boolean {
			return !str.includes(char);
		}, `no character of '${str}'`);
	}

	/**
	 * Returns a parser that checks every string in `strings` and yields the first matching string.
	 *
	 * @param strings
	 */
	static oneStringOf(strings: readonly string[]): Parser<string> {
		return this.or(...strings.map(x => this.string(x))).describe(strings.map(x => `'${x}'`).join(' or '));
	}

	/**
	 * Returns a parser that matches as many characters as it can, as long as they are included in `str`.
	 * `P.manyOf('ab')` is the same and more performant as `P.oneOf('ab').many().map(x => x.join(''))`.
	 *
	 * @param str
	 */
	static manyOf(str: string): Parser<string> {
		return new Parser<string>(function _manyOf(context): InternalParseResult<string> {
			let i = context.position;
			for (; i < context.input.length; i++) {
				if (!str.includes(context.input[i])) {
					break;
				}
			}
			return context.succeedAt(i, context.sliceTo(i));
		});
	}

	/**
	 * Returns a parser that matches anything until a character included in `str` is encountered.
	 * `P.manyNotOf('ab')` is the same and more performant as `P.noneOf('ab').many().map(x => x.join(''))`.
	 *
	 * @example A simple string without escape characters.
	 * P.manyNotOf('"').trim(P.string('"'))
	 *
	 * @param str
	 */
	static manyNotOf(str: string): Parser<string> {
		return new Parser<string>(function _manyOf(context): InternalParseResult<string> {
			let i = context.position;
			for (; i < context.input.length; i++) {
				if (str.includes(context.input[i])) {
					break;
				}
			}
			return context.succeedAt(i, context.sliceTo(i));
		});
	}

	/**
	 * Create a custom parser from a parsing function.
	 *
	 * @param parsingFunction
	 */
	static custom<SType extends STypeBase>(parsingFunction: ParseFunction<SType>): Parser<SType> {
		return new Parser(parsingFunction);
	}

	/**
	 * Returns a parser that matches and yields a character, which char code is between the char codes of `begin` and `end`.
	 *
	 * @param begin
	 * @param end
	 */
	static range(begin: string, end: string): Parser<string> {
		const beginCharCode = begin.charCodeAt(0);
		const endCharCode = end.charCodeAt(0);
		return P_HELPERS.testCharCode(function _range(charCode): boolean {
			return beginCharCode <= charCode && charCode <= endCharCode;
		}, `${begin}-${end}`);
	}

	/**
	 * Returns a parser that matched until `fn` returns false. Then yields all matched characters as a string.
	 *
	 * @param fn
	 */
	static takeWhile(fn: (char: string) => boolean): Parser<string> {
		return new Parser(function _takeWhile(context): InternalParseResult<string> {
			let endIndex = context.position;
			while (endIndex < context.input.length && fn(context.input[endIndex])) {
				endIndex++;
			}
			return context.succeedAt(endIndex, context.input.slice(context.position, endIndex));
		});
	}

	/**
	 * Wraps a parser, allowing for recursion. `fn` is only called when the parser is in use.
	 * Recursion can easily lead to infinite loops.
	 *
	 * @param fn
	 */
	static reference<SType extends STypeBase>(fn: () => Parser<SType>): ParserRef<SType> {
		return new Parser<SType>(function _reference(context): InternalParseResult<SType> {
			return fn().p(context);
		});
	}
}
