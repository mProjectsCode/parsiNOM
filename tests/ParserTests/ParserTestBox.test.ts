import { P } from '../../src/ParsiNOM';
import { ParserTestData, testParser, testParserAdvanced } from '../TestHelpers';
import { describe } from 'bun:test';

describe.each([
	['', false],
	['this', true],
	['foo', false],
])(`box`, (str, shouldSucceed) => {
	const parser = P.string('this').box('some error');
	testParser(parser, str, shouldSucceed);
});

describe.each<ParserTestData<string>>([
	{
		input: '',
		shouldSucceed: false,
		furthest: 0,
		expected: [`('this' as part of something)`],
	},
	{
		input: 'foo',
		shouldSucceed: false,
		furthest: 0,
		expected: [`('this' as part of something)`],
	},
	{
		input: 'this',
		shouldSucceed: true,
		ast: 'this',
		toIndex: 4,
	},
])(`box advanced`, data => {
	const parser = P.string('this').box('something');
	testParserAdvanced(parser, data);
});

describe.each<ParserTestData<string[]>>([
	{
		input: '',
		shouldSucceed: false,
		furthest: 0,
		expected: [`('this' as part of something)`],
	},
	{
		input: 'this',
		shouldSucceed: false,
		furthest: 4,
		expected: [`('that' as part of something)`],
	},
	{
		input: 'that',
		shouldSucceed: false,
		furthest: 0,
		expected: [`('this' as part of something)`],
	},
	{
		input: 'thisthat',
		shouldSucceed: true,
		ast: ['this', 'that'],
		toIndex: 8,
	},
	{
		input: 'thisthatfoo',
		shouldSucceed: true,
		ast: ['this', 'that'],
		toIndex: 8,
	},
	{
		input: 'foo',
		shouldSucceed: false,
		furthest: 0,
		expected: [`('this' as part of something)`],
	},
])(`box sequence advanced`, data => {
	const parser = P.sequence(P.string('this'), P.string('that')).box('something');
	testParserAdvanced(parser, data);
});

describe.each<ParserTestData<[[string[], string] | undefined, string]>>([
	{
		input: 'aaaaa',
		shouldSucceed: false,
		furthest: 5,
		expected: [`('a' as part of many 'a's)`, "('b' as part of the end of the 'a's)"],
	},
	{
		input: 'aab',
		shouldSucceed: false,
		furthest: 3,
		expected: ["'c'"],
	},
	{
		input: 'c',
		shouldSucceed: true,
		ast: [undefined, 'c'],
		toIndex: 1,
	},
	{
		input: 'aabc',
		shouldSucceed: true,
		ast: [[['a', 'a'], 'b'], 'c'],
		toIndex: 4,
	},
])(`box multiple errors advanced`, data => {
	const parser = P.sequence(P.sequence(P.string('a').many().box("many 'a's"), P.string('b').box("the end of the 'a's")).optional(), P.string('c'));
	testParserAdvanced(parser, data);
});
