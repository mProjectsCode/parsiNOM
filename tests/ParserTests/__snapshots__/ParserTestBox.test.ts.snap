// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`box '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "('this' as part of some error)",
  ],
  "pos": 0,
}
`;

exports[`box 'this' conditional success tests AST matches snapshot 1`] = `"this"`;

exports[`box 'foo' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "('this' as part of some error)",
  ],
  "pos": 0,
}
`;
